### PR TITLE
Support PyTrees as model outputs in test infra

### PR DESCRIPTION
**Pull request by sgligorijevicTT**
### Ticket
N/A

### Problem description
Some model implementations, primarily from huggingface, return classes with multiple fields as outputs instead of a single jax Array. This causes type errors during comparisons. It is also possible other models might want to have multiple outputs.

### What's changed
This PR changes the test infrastructure to work with arbitrary PyTrees instead of a single tensor, using tree utilities from jax. That includes any custom class a model wants to return as long as it's properly registered as a PyTree node(which includes HF classes).

### Checklist
- [X] New/Existing tests provide coverage for changes
- [X] Update type annotations where appropriate

### CODE REVIEW

File: tests/infra/base_tester.py [hunk 1]
```diff
@@ -67,7 +67,9 @@ def _match_data_types(self, *tensors: Tensor) -> Sequence[Tensor]:
...
-        return [
-            tensor.astype("float32") if tensor.dtype.str != "float32" else tensor
-            for tensor in tensors
-        ]
+        return jax.tree.map(
+            lambda tensor: tensor.astype("float32")
+            if tensor.dtype.str != "float32"
+            else tensor,
+            tensors,
+        )
```
Comment (kmitrovicTT) [inline tests/infra/base_tester.py:75]: What is the return type of this?

File: tests/infra/comparison.py [hunk 2]
```diff
@@ -74,26 +74,21 @@ def disable_all(self) -> None:
...
-    assert isinstance(device_output, jax.Array) and isinstance(
-        golden_output, jax.Array
-    ), f"Currently only jax.Array is supported"
-
-    eq = (device_output == golden_output).all()
-
-    assert eq, f"Equal comparison failed."
+def compare_equal(device_output: PyTree, golden_output: PyTree) -> None:
+    passed = jax.tree.map(lambda x, y: (x == y).all(), device_output, golden_output)
+    assert jax.tree.all(passed), f"Equal comparison failed."
```
Comment (kmitrovicTT) [inline tests/infra/comparison.py:79]: Take note that is now heavily coupled to jax. `Tensor` alias served as a way to abstract the concrete framework underneath. Now that you use `Pytree`, it is solely a jax thing. Even asserts have been removed. Same for functions below.

This `Pytree` structure is so weird to me. How can `device_output` be a `Pytree`? It is a tensor by design.

Comment (sgligorijevicTT) [reply tests/infra/comparison.py:79]: > How can `device_output` be a `Pytree`? It is a tensor by design.

This caught me off guard too.
Models can return anything, and specifically huggingface classes take full advantage of that by returning a custom class( example: [FlaxMaskedLMOutput](https://huggingface.co/docs/transformers/v4.49.0/en/main_classes/output#transformers.modeling_flax_outputs.FlaxMaskedLMOutput) is returned by [FlaxAlbertForMaskedLM](https://huggingface.co/docs/transformers/en/model_doc/albert#transformers.FlaxAlbertForMaskedLM.__call__) ). This is so they can return multiple tensors(for example, specifying `output_hidden_states=True` to a model causes it to also return outputs of every transformer block alongside the logits).

Regarding generality across frameworks, `jax.tree` functions *should* be agnostic to what exactly is inside the PyTree, and as long as the lambda passed to some `jax.tree` function can handle the leaf nodes everything should work. This code is also reusable in cases other frameworks return standard containers of custom tensor types. Main point of concern would be if models from other frameworks start returning custom dataclasses too.

File: tests/infra/base_tester.py (hunk 1 above)
Comment (sgligorijevicTT) [reply tests/infra/base_tester.py:75]: `tree.map` return a pytree with identical structure as the input pytree except with all leaf nodes replaced by the output type of the lambda.

Comment (kmitrovicTT) [reply tests/infra/base_tester.py:75]: So is the return type hint of this function ok?

File: tests/infra/comparison.py (hunk 2 above)
Comment (kmitrovicTT) [reply tests/infra/comparison.py:79]: We'll deal with it if that issue arises.

**Review by kmitrovicTT — APPROVED** [review]
